### PR TITLE
feat(routing): add invalidate window as a service for SR based routing

### PR DESCRIPTION
**Pull request by Aprabhat19**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
add invalidate window as a service for SR based routing

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
 
 - Hit the route, to update the config
 ```
 curl --location --request PATCH 'http://localhost:8080/account/merchant_1730123706/business_profile/pro_TtvgDfAdgudrReoeeUJb/dynamic_routing/success_based/config/routing_lmPpwoTevEaYMVogesHs' \
--header 'Content-Type: application/json' \
--header 'api-key: test_admin' \
--data '{
  "params": [
    "Currency"
  ],
  "config": {
    "min_aggregates_size": 8,
    "current_block_threshold": {
        "max_total_count": 22
    }
  }
}'
```
-  The old config would get invalidated
Screenshot 2024-10-28 at 11 25 13 PM

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by Chethan-rao — CHANGES_REQUESTED** [review]
Invoke `invalidate_config_window` in `update_success_rate` because once configs are updated, all the metrics in dynamo has to be refreshed(invalidated)

**Review by Chethan-rao — CHANGES_REQUESTED** [review]

File: crates/router/src/core/routing.rs [hunk 1]
```diff
@@ -1445,6 +1446,18 @@ pub async fn success_based_routing_update_configs(
...
+    let dynamic_routing_id = helpers::generate_tenant_business_profile_id(
+        &state.tenant.redis_key_prefix,
+        profile_id.get_string_repr(),
+    );
+    state
+        .grpc_client
+        .dynamic_routing
+        .success_rate_client
+        .as_ref()
+        .map(|a| a.invalidate_config_window(dynamic_routing_id));
```
Comment (Chethan-rao) [inline crates/router/src/core/routing.rs:1459]: Isn't this an async function? you'll have to await it and handle error too

**Review by Chethan-rao — DISMISSED** [review]

**Review by Chethan-rao — DISMISSED** [review]

File: config/development.toml [hunk 2]
```diff
@@ -766,3 +766,7 @@ card_networks = "Visa, AmericanExpress, Mastercard"
 
 [network_tokenization_supported_connectors]
 connector_list = "cybersource"
+
+[grpc_client.dynamic_routing_client]
```
Comment (NishantJoshi00) [inline config/development.toml:770]: can you also add it in `example.toml`?

Also, do we need to have this here in `development.toml`

File: crates/external_services/src/grpc_client/dynamic_routing.rs [hunk 3]
```diff
@@ -209,6 +215,24 @@ impl SuccessBasedDynamicRouting for SuccessRateCalculatorClient<Channel> {
...
         Ok(response)
     }
+
+    async fn invalidate_success_rate_routing_keys(
+        &self,
+        id: String,
+    ) -> DynamicRoutingResult<InvalidateWindowsResponse> {
+        let request = tonic::Request::new(InvalidateWindowsRequest { id });
+
+        let mut client = self.clone();
```
Comment (NishantJoshi00) [inline crates/external_services/src/grpc_client/dynamic_routing.rs:225]: Why are we cloning here?

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

File: config/development.toml (hunk 2 above)
Comment (Aprabhat19) [reply config/development.toml:770]: Already there

File: crates/external_services/src/grpc_client/dynamic_routing.rs (hunk 3 above)
Comment (Aprabhat19) [reply crates/external_services/src/grpc_client/dynamic_routing.rs:225]: Cause we need a mutable ref to the Channel , and also cloning the channel is cheap

File: crates/router/src/core/routing.rs [hunk 4]
```diff
@@ -1445,6 +1448,27 @@ pub async fn success_based_routing_update_configs(
...
+        profile_id.get_string_repr(),
+    );
+    state
+        .grpc_client
+        .dynamic_routing
+        .success_rate_client
+        .as_ref()
+        .async_map(|sr_client| async {
+            sr_client
+                .invalidate_success_rate_routing_keys(prefix_of_dynamic_routing_keys)
```
Comment (NishantJoshi00) [inline crates/router/src/core/routing.rs:1463]: we should try and use that wildcard graph approach once done.

cc: @tinu-hareesswar

**Review by NishantJoshi00 — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Pull request merged**